### PR TITLE
Remove shipments schema

### PR DESCRIPTION
**Pull request by mbj**
Post extraction from #398

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb Ready for review after #398

File: backend/app/views/spree/admin/orders/_line_items.html.erb [hunk 1]
```diff
@@ -50,13 +50,13 @@
...
-      <% order.shipments.select(&:selected_shipping_rate).each do |shipment| %>
+      <% if order.shipping_method %>
         <tr class="show-method total">
           <td colspan="4">
-            <strong><%= shipment.selected_shipping_rate.name %></strong>
+            <strong><%= order.shipping_method.name %></strong>
           </td>
           <td class="total" align="center">
-            <span><%= shipment.display_cost %></span>
+            <span><%= Spree::Money.new(order.shipment_total, currency: order.currency).to_html %></span>
```
Comment (dkubb) [inline backend/app/views/spree/admin/orders/_line_items.html.erb]: Can this be replaced with `Spree::Order#display_shipment_total` ?

File: core/app/views/spree/order_mailer/confirm_email.text.erb [hunk 2]
```diff
@@ -18,9 +18,7 @@
   <% end %>
 <% end %>
 
-<% @order.shipments.group_by { |s| s.selected_shipping_rate.try(:name) }.each do |name, shipments| %>
-<%= Spree.t(:shipping) %>: <%= name %> <%= Spree::Money.new(shipments.sum(&:cost), currency: @order.currency) %>
-<% end %>
+<%= Spree.t(:shipping) %>: <%= @order.shipping_method.name %> <%= Spree::Money.new(@order.shipment_total, currency: @order.currency) %>
```
Comment (dkubb) [inline core/app/views/spree/order_mailer/confirm_email.text.erb]: This can use `Spree::Order#display_shipment_total`.

File: backend/app/controllers/spree/admin/line_items_controller.rb [hunk 3]
```diff
@@ -37,7 +37,13 @@ def destroy
       # @return [undefined]
       #
       def update_quantity(quantity)
-        parent.contents.update_cart(update_attributes(quantity))
+        order = parent
+        order.contents.update_cart(update_attributes(quantity))
+
+        # Shipments totals are normally not updated in completed state
+        # Hence we need to trigger this explicitly here.
+        order.refresh_shipping
```
Comment (dkubb) [inline backend/app/controllers/spree/admin/line_items_controller.rb]: Could everything in this method, aside from the `order.refresh_shipping` line be extracted into it's own commit?

File: backend/spec/controllers/spree/admin/line_items_controller_spec.rb [hunk 4]
```diff
@@ -57,6 +59,9 @@ def do_request
               }
             }
           )
+
+        expect(order).to have_received(:refresh_shipping)
+        expect(order).to have_received(:save)
```
Comment (dkubb) [inline backend/spec/controllers/spree/admin/line_items_controller_spec.rb]: I think this line could be extracted early.

File: backend/spec/controllers/spree/admin/line_items_controller_spec.rb [hunk 5]
```diff
@@ -110,6 +115,9 @@ def do_request
             }
           }
         )
+
+      expect(order).to have_received(:refresh_shipping)
+      expect(order).to have_received(:save)
```
Comment (dkubb) [inline backend/spec/controllers/spree/admin/line_items_controller_spec.rb]: I think this line could be extracted early.

File: core/app/models/spree/order.rb [hunk 6]
```diff
@@ -48,9 +47,19 @@ class Order < Spree::Base
       end
     end
 
-    ORDER_STATES = IceNine.deep_freeze(state_machine.states.map { |state| state.name.to_s }.to_set)
+    STATES_REQUIRE_SHIPMENT_METHOD =
+      IceNine.deep_freeze(%w[delivery payment complete canceled resumed].to_set)
+
+    private_constant(*constants(false) - %i[NUMBER_PATTERN])
```
Comment (dkubb) [inline core/app/models/spree/order.rb]: This line, and the 3 statements that follow could be extracted into their own commit. I know it's tiny and probably inconsequential, but in this area only the `STATES_REQUIRE_SHIPMENT_METHOD` needs to be part of the shipment removal commit(s).

File: core/app/models/spree/order.rb [hunk 7]
```diff
@@ -431,13 +429,110 @@ def quantity
...
     def require_email
       !(cart? || address?)
     end
 
+    # Test if shipping method is required
+    #
+    # @return [Boolean]
+    def require_shipping_method?
+      STATES_REQUIRE_SHIPMENT_METHOD.include?(state)
+    end
```
Comment (dkubb) [inline core/app/models/spree/order.rb]: You could release this and the `STATES_REQUIRE_SHIPMENT_METHOD` constant first, as well as the validation that asserts shipment_total is $0 when `#require_shipping_method?` returns false.

File: core/app/models/spree/order_updater.rb [hunk 8]
```diff
@@ -33,28 +32,19 @@ def update_payment_state
...
   private
 
-    # Updates the following Order total values:
+    # Updates the order total values
     #
-    # +payment_total+      The total value of all finalized Payments (NOTE: non-finalized Payments are excluded)
-    # +item_total+         The total value of all LineItems
-    # +adjustment_total+   The total value of all adjustments (promotions, credits, etc.)
-    # +promo_total+        The total value of all promotion adjustments
-    # +total+              The so-called "order total."  This is equivalent to +item_total+ plus +adjustment_total+.
```
Comment (dkubb) [inline core/app/models/spree/order_updater.rb]: You could probably update the comments here in an early commit.

File: core/app/models/spree/shipping_rate_estimator.rb [hunk 9]
```diff
@@ -55,21 +58,23 @@ def package
...
     # Calculate a rate for shipping method
     #
     # @param [Spree::ShippingMethod] shipping_method
     #
-    # @return [Float]
-    #   yes float is sadly the return type currently, blame spree_active_shipping
+    # @return [BigDecimal]
     def calculate(shipping_method)
-      shipping_method.calculator.compute(package)
+      shipping_method.calculator.compute(package).to_d # Yes spree_active_shipping returns a float
```
Comment (dkubb) [inline core/app/models/spree/shipping_rate_estimator.rb]: Can this be extracted into an early commit? I assume it won't cause any problems to return `BigDecimal` from here instead of a `Float`?

File: core/spec/mailers/order_mailer_spec.rb [hunk 10]
```diff
@@ -5,7 +5,8 @@
   include EmailSpec::Matchers
 
   let(:order) do
-    order = create(:order)
+    order = create(:completed_order, shipment_total: 10)
```
Comment (dkubb) [inline core/spec/mailers/order_mailer_spec.rb]: I noticed that the `shipment_total` is now specified in several specs where it wasn't previously. What specific change prompted you to add this? Can be be extracted into it's own commit? It seems like it'd reduce a large amount of the diff around the specs themselves.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 11]
```diff
@@ -169,6 +174,32 @@ def persist_user_address
...
+    #
+    # @api private
+    #
+    # @return [undefined]
+    def update_shipping
+      case @order.state
+      when *ASSIGN_SHIPPING_STATES
+        assign_shipping_method_from_params
+      when *RESET_SHIPPING_STATES
+        @order.assign_shipping_method(nil)
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: I notice there are two places where you do `assign_shipping_method(nil)`. Should a method named `#remove_shipping_method` be created to do this?

Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Would it be possible to refactor this case statement into a Hash that maps the states to the method to execute, with the default being an `id` method that is a noop?

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 12]
```diff
@@ -169,6 +174,32 @@ def persist_user_address
...
+    end
+
+    # Assign shipping method from params
+    #
+    # @return [undefined]
+    #
+    # @api private
+    def assign_shipping_method_from_params
+      shipping_method_id = params.dig(:order, :shipping_method_id) or return
+      shipping_method = ShippingMethod.find_by(id: shipping_method_id) or return
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: What about using `and` instead of this approach?

File: frontend/app/views/spree/checkout/_summary.html.erb [hunk 13]
```diff
@@ -18,12 +18,10 @@
...
 
-    <% if order.shipments.any? %>
-      <tr data-hook="shipping_total">
-        <td><%= Spree.t(:shipping_total) %>:</td>
-        <td><%= Spree::Money.new(order.shipments.to_a.sum(&:cost), currency: order.currency).to_html %></td>
-      </tr>
-    <% end %>
+    <tr data-hook="shipping_total">
+      <td><%= Spree.t(:shipping_total) %>:</td>
+      <td><%= Spree::Money.new(order.shipment_total, currency: order.currency).to_html %></td>
```
Comment (dkubb) [inline frontend/app/views/spree/checkout/_summary.html.erb]: This could use `Spree::Order#display_shipment_total`.

File: frontend/app/views/spree/orders/_adjustments.html.erb [hunk 14]
```diff
@@ -5,11 +5,11 @@
...
   <% end %>
 
-  <% @order.shipments.each do |shipment| %>
+  <% if @order.shipping_method %>
     <tr>
-      <td colspan="4" align='right'><h5><%= Spree.t(:shipping) %>: <%= shipment.shipping_method.name %></h5></td>
+      <td colspan="4" align='right'><h5><%= Spree.t(:shipping) %>: <%= @order.shipping_method.name %></h5></td>
       <td colspan='2'>
-        <h5><%= shipment.display_cost %></h5>
+        <h5><%= @order.shipment_total %></h5>
```
Comment (dkubb) [inline frontend/app/views/spree/orders/_adjustments.html.erb]: Should this use `Spree::Order#display_shipment_total` instead of the raw method?

File: frontend/app/views/spree/shared/_order_details.html.erb [hunk 15]
```diff
@@ -92,12 +80,10 @@
...
   <tfoot id='shipment-total'>
-    <% order.shipments.group_by { |s| s.selected_shipping_rate.name }.each do |name, shipments| %>
-      <tr class="total" data-hook='shipment-row'>
-        <td colspan="4"><%= Spree.t(:shipping) %>: <strong><%= name %></strong></td>
-        <td class="total"><span><%= Spree::Money.new(shipments.sum(&:cost), currency: order.currency).to_html %></span></td>
-      </tr>
-    <% end %>
+    <tr class="total" data-hook="shipment-row">
+      <td colspan="4"><%= Spree.t(:shipping) %>: <strong><%= order.shipping_method.name %></strong></td>
+      <td class="total"><span><%= Spree::Money.new(order.shipment_total, currency: order.currency).to_html %></span></td>
```
Comment (dkubb) [inline frontend/app/views/spree/shared/_order_details.html.erb]: Can this use `Spree::Order#display_shipment_total`.

File: core/app/models/spree/order.rb (hunk 7 above)
Comment (mbj) [inline core/app/models/spree/order.rb]: We had that one before, and I'm happy to state it again: This validation is not guaranteed before this branch, it would fail very often and fixing it inside the current model would be more work than changing the model.

File: core/app/models/spree/shipping_rate_estimator.rb (hunk 9 above)
Comment (mbj) [inline core/app/models/spree/shipping_rate_estimator.rb]: I've to reject this as the explicit coercion is only needed when the silently coercing `Spree::ShippingRate` model is gone.

File: core/spec/mailers/order_mailer_spec.rb (hunk 10 above)
Comment (mbj) [inline core/spec/mailers/order_mailer_spec.rb]: This is likely an artifact from an intermediary state, it can possibly be removed. I'll check each instance.

File: backend/app/controllers/spree/admin/line_items_controller.rb (hunk 3 above)
Comment (mbj) [inline backend/app/controllers/spree/admin/line_items_controller.rb]: No because its dependent on a writable `Order#shipping_method`. and the shipments are implicitly refeshed on any interactions with the order right now, regardless of completion state.

File: core/spec/mailers/order_mailer_spec.rb (hunk 10 above)
Comment (mbj) [inline core/spec/mailers/order_mailer_spec.rb]: In many cases, such as the order specs we do need to explicitly set the `shipment_total` to a known value that is in-line with the rate estimations (or explicitly not to test the stale rate code path). This case was an artifact and could be removed.

**Comment by dkubb** [comment]
LGTM. Will merge when CI is green.
